llvm/lib/Transforms/Utils/LoopIterCountsToAssumptions.cpp

```diff
@@ -0,0 +1,191 @@
+//===-- LoopIterCountsToAssumptions.cpp - add Loop assumes -*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/LoopIterCountsToAssumptions.h"
+#include "llvm/Analysis/AliasAnalysis.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/Analysis/MemorySSA.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Transforms/Utils/LoopUtils.h"
+#include "llvm/Transforms/Utils/ScalarEvolutionExpander.h"
+
+#define DEBUG_TYPE "loop-iter-count-assumes"
+
+using namespace llvm;
+namespace {
+
+/// get AddRec scalar evolution, if it can be found in Op
+const SCEVAddRecExpr *getAddRecSCEV(Value *Op, ScalarEvolution *SE) {
+  if (!SE->isSCEVable(Op->getType()))
+    return nullptr;
+
+  return dyn_cast<SCEVAddRecExpr>(SE->getSCEV(Op));
+}
+
+/// get the SCEV of the loop IV
+std::pair<const SCEVAddRecExpr *, int>
+getLoopCmpIVAddRecExpr(ICmpInst &LoopCmpInstr, ScalarEvolution *SE) {
+  std::pair<const SCEVAddRecExpr *, int> BundledIV(nullptr, -1);
+
+  const SCEVAddRecExpr *S0 = getAddRecSCEV(LoopCmpInstr.getOperand(0), SE);
+  const SCEVAddRecExpr *S1 = getAddRecSCEV(LoopCmpInstr.getOperand(1), SE);
+
+  if (S0 && S1) {
+    LLVM_DEBUG(
+        dbgs() << "LoopIterCountsToAssumptions-Warning: Encountered two ADDRec "
+                  "expressions in the Loop Compare instruction. Currently it "
+                  " is expected that the compare consists of the IV and a "
+                  "static bound, that is loop invariant.");
+    return BundledIV;
+  }
+
+  if (S0) {
```

Comment:
Maybe we can:
```
std::pair<const SCEVAddRecExpr *, int> BundledIV(S0, 0);
if (S1) {
   BundledIV.first = S1;
    BundledIV.second = 1;
}
return BundledIV;

Adjusting the first if to return null and -1 pair directly.....
```

Reply:
we still miss the case, if S0 is not the SCEVAddRecExpr

Reply:
And we don't care about the second value when the SCEV is nullptr....

Reply:
In the end I do not go with that option, because I find `std::optional` nicer

```
  std::optional<std::pair<const SCEVAddRecExpr *, int>> LoopBoundIV =
      getLoopCmpIVAddRecExpr(*LoopCmpInstr, SE);
  if (!LoopBoundIV) {
  // ....
    return PreservedAnalyses::all();
  }
  auto [AddRecExpr, IVOperand] = *LoopBoundIV;
```